[UserStory] QT4BJ extension has its own panel in BlueJ Preferences>Extensions #112

holsappled opened the pull request:
Created Preferences class and associated acceptance test as an AcceptanceTest.md.

holsappled committed:
added Preferences class

holsappled committed:
Create AcceptanceTest.md

jody commented on AcceptanceTest.md:
```diff
@@ -0,0 +1,23 @@
+# AcceptanceTest for QualityToolsExtension.jar
```
Is there **exactly one and only one** acceptance test?  Is a new file needed for an additional test?

Is the intent that this only apply to a file named _QualityToolsExtension.jar_?  Does it need to be rewritten or replaced if the name of the jar file evolves during the lifetime of the product?

shadycloud replied:
1) Within the ```AcceptanceTest.md``` file.  There are two acceptance tests.  First test is to verify ```QualityTools``` Extension displays.  The second acceptance test being Quality Tools Extension displays from Preferences “**Extensions Tab**”.

2) It seems our organization has decided to implement that name for our production jar files as in User story #149 and #138 which could be confusing. We are proposing that we name it ```ProductAcceptanceTest.md```

shadycloud replied:
It seems the MetroCS organization plans on the future implementation of semantic versioning in our production jar files as detailed in User story #149 and #138. However, slight modifications will need to be updated ``` "rewritten or replaced" ``` in the file name and body content until a more permanent strategy can be developed to automagically handle semantic versioning of file names. https://semver.org/

jody commented on AcceptanceTest.md:
```diff
@@ -0,0 +1,23 @@
+# AcceptanceTest for QualityToolsExtension.jar
+
+Due to the difficulty of JUnit tests developers created the following acceptance test to verify the start up functionality of the BlueJ application. 
```
What is meant by "Due to the difficulty of JUnit tests"?  What is the intent of the opening phrase of this document?

holsappled replied:
Invoking the BlueJ proxy object to create a JUnit test.

jody commented on AcceptanceTest.md:
```diff
@@ -0,0 +1,23 @@
+
+# Steps 
+
+1) Gradle build the project with JDK version 11
```
As I progress through this document, I begin to understand that this is intended more as a "one off" acceptance criterion than something useful for longer duration of product development and maintenance.

If this is the case, then it should appear as "Acceptance Criteria" in the associated issue, rather than as a document in the repository.

Consider copying the verification steps to the appropriate issue(s) and deleting this document.

holsappled replied:
After discussing with the group, having a JTextField displayed in the quality tools extension tab serves no purpose. The intent was to match the #112 user story "picture display"

jody commented on src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java:
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
+class Preferences implements PreferenceGenerator
+{
+    private JPanel myPanel;
+    private JTextField color;
```
What is the intent of this field?  It seems irrelevant.

holsappled replied:
Field is needed to create JTextField from the Swing API.  

After discussing with the group, having a JTextField displayed in the quality tools extension tab serves no purpose.  The intent was to match the #112 user story "picture display"

new commit will remove field

jody replied:
> The intent was to match the #112 user story "picture display"

Matching that with something relevant to the product, such as, "Quality Analysis Tools", would have been fine.  "color" just seems irrelevant.

jody commented on src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java:
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
+ */
+class Preferences implements PreferenceGenerator
+{
+    private JPanel myPanel;
```
All fields and methods should be commented, using Javadoc-style comments.

jody commented on src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java:
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
 }
+
+/**
+ * Create a Preference Panel located in the BlueJ application
```
This comment seems to indicate that this class determines the location of the panel.  That doesn't seem to be the actual way this is working.

jody commented on src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java:
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
+        myPanel = new JPanel();
+        myPanel.add (new JLabel ("Quality Analysis Tools"));
+        color = new JTextField (40);
+        myPanel.add (color);
```
What is "color"?

holsappled replied:
We will remove "color" object as it was used to store the JTextField.  There is no user input required for this user story.  Will remove lines of code associated with "color".  

For reference and to answer your question see the information below.  

`color = new JTextField (40);`
API recommends Note: 
"We encourage you to specify the number of columns for each text field. If you do not specify the number of columns or a preferred size, then the field's preferred size changes whenever the text changes, which can result in unwanted layout updates."

https://docs.oracle.com/javase/tutorial/uiswing/components/textfield.html

jody replied:
Just a note that it is less about "user input" and more about the choice of "color" in particular as having no obvious relevance and likely being an unintended artifact of a copy-paste of work done by others.

jody commented on src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java:
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
+        myPanel.add (new JLabel ("Quality Analysis Tools"));
+        color = new JTextField (40);
+        myPanel.add (color);
+        // Load the default value
```
What is meant by "default value"?

jody reviewed (changes requested):
Please have multiple members of your team perform a code review and attest to the quality and conformance to conventions.

holsappled committed:
Delete AcceptanceTest.md

Deleted AcceptanceTest.md.  Content of AcceptanceTest.md was better suited inside user story #112

holsappled committed:
Updated changes from code review pull request

Added Javadoc-style comments, deleted example JPanel.

holsappled commented:
New commits will reflect BlueJ application under BlueJ->Preferences->Extensions Tab

<img width="799" alt="Screen Shot 2019-12-04 at 6 38 37 AM" src="https://user-images.githubusercontent.com/10522689/70147765-2b764a00-1662-11ea-9f13-41c17c529da2.png">

jody reviewed (changes requested):
This code doesn't appear to follow the coding convention guidelines for the project.
See: [Coding Convention Guide](https://github.com/MetroCS/QualityToolsForBlueJ/blob/master/DevelopmentConventions.md#Coding-Convention-Guide)

holsappled committed:
Added comments to follow conventions

Method headers, class headers, authors

jody reviewed (approved):
Will approve to close this out. Need to open bug report to fix formatting (misplaced '{', etc.) to adhere to coding convention.
